sunpy/map/mapbase.py

```diff
@@ -1963,50 +1963,122 @@ def draw_grid(self, axes=None, grid_spacing: u.deg = 15*u.deg, annotate=True, **
+
+        To avoid trigger Matplotlib auto-scaling, these patches are added as generic
+        artists instead of patches.  One consequence is that the plot legend is not
+        populated automatically when the limb is specified with a text label.
         """
         # Put import here to reduce sunpy.map import time
         from matplotlib import patches
+        from matplotlib.path import Path
 
         # Don't use _check_axes() here, as drawing the limb works fine on none-WCSAxes,
         # even if the image is rotated relative to the axes
         if not axes:
             axes = wcsaxes_compat.gca_wcs(self.wcs)
+        is_wcsaxes = wcsaxes_compat.is_wcsaxes(axes)
 
-        transform = wcsaxes_compat.get_world_transform(axes)
-        if wcsaxes_compat.is_wcsaxes(axes):
-            radius = self.rsun_obs.to(u.deg).value
-        else:
-            radius = self.rsun_obs.value
-        c_kw = {'radius': radius,
-                'fill': False,
+        c_kw = {'fill': False,
                 'color': 'white',
-                'zorder': 100,
-                'transform': transform
-                }
+                'zorder': 100}
         c_kw.update(kwargs)
 
-        circ = patches.Circle([0, 0], **c_kw)
-        axes.add_artist(circ)
+        # Obtain the solar radius and the world->pixel transform
+        if not is_wcsaxes:
+            radius = self.rsun_obs.value
+            transform = axes.transData
+        else:
+            radius = self.rsun_obs.to_value(u.deg)
 
-        return [circ]
+            if isinstance(self.coordinate_frame, Helioprojective):
+                frame = self.coordinate_frame
+            else:
+                frame = Helioprojective(observer=self.observer_coordinate,
+                                        obstime=self.date, rsun=self.rsun_meters)
+            transform = axes.get_transform(frame)
+
+        # transform is always passed on as a keyword argument
+        c_kw.setdefault('transform', transform)
+
+        # If not WCSAxes or if the map's frame matches the axes's frame, we can use Circle
+        if not is_wcsaxes or frame == axes._transform_pixel2world.frame_out:
+            c_kw.setdefault('radius', radius)
+
+            circ = patches.Circle([0, 0], **c_kw)
+            axes.add_artist(circ)
+
+            return circ, None
+
+        # Otherwise, we use Polygon to be able to distort the limb
+
+        theta = np.linspace(0, 2*np.pi, resolution+1)[:-1] << u.rad
+        # Use the Cartesian approximation of helioprojective coordinates because the Sun is small
+        Tx, Ty = radius * np.cos(theta), radius * np.sin(theta)
+
+        # Perform the round-trip transformation world->pixel->world
+        true_vertices = np.array([Tx, Ty]).T
+        apparent_vertices = transform.inverted().transform(transform.transform(true_vertices))
+
+        # Points that move by less than 1 arcmin are intepreted to be on the "near" side of the Sun
```

Comment:
I'm struggling to get how this logic works. If I'm plotting the SDO limb from the POV of STEREO, at a 45 deg observer difference, don't all the points move? I think I'm probably not understanding exactly what the transform chain above is doing.

Reply:
See https://github.com/sunpy/sunpy/pull/5237#discussion_r618488272

The load-bearing step is `axes_world(2D)`->`axes_world(3D)`.  If the initial point is on the far side of the Sun for the axes's observer, this step will not be the inverse of the earlier `axes_world(3D)`->`axes_world(2D)`, and thus the round-trip transformation fails.